### ISSUE TITLE
Detailed form errors

### ISSUE THREAD

**Issue by lucascourot** [FriendsOfSymfony/FOSRestBundle#842]
At the moment the form errors look something like:

```javascript
{
  "code": 400,
  "message": "Validation Failed",
  "errors": {
    "children": {
      "username": {
        "errors": [
          "This value should not be blank."
        ]
      }
    }
  }
}
```

I would like to have both error message and error code for each validated form field, so the json would look like this example below http://www.vinaysahni.com/best-practices-for-a-pragmatic-restful-api#errors

```javascript
{
  "code" : 1024,
  "message" : "Validation Failed",
  "errors" : [
    {
      "code" : 5432,
      "field" : "first_name",
      "message" : "First name cannot have fancy characters"
    },
    {
       "code" : 5622,
       "field" : "password",
       "message" : "Password cannot be blank"
    }
  ]
}
```

Is it possible to obtain this with Symfony? Is it possible to associate an error code to a validation constraint?  Or do you see any other solution?

**Comment by broncha** [comment]
+1 for consistency between hash and array

**Comment by renatomefidf** [comment]
@lsmith77 How can I return errors from a different kind of exception? Which is not form errors.
I can see all the exceptions being turned into a FlattenException, which means my custom exceptions are not available at ExceptionHandler/Wrapper. How can I achieve custom errors?

Thank you!

**Comment by eXtreme** [comment]
@renatomefidf I'm struggling with the same thing. I tried writing my ExceptionWrapperHandler but there is no original exception available to retrieve additional data from it. Has anyone managed to do this? I have a custom exception which carries command validation errors and I can't find any way to show them within fos rest.

**Comment by Ener-Getick** [comment]
@lsmith77 what do you think about using a custom exception listener with an exception normalizer (something like [the symfony normalizers](https://github.com/symfony/symfony/tree/2.8/src/Symfony/Component/Serializer/Normalizer)) which would convert the exception to an array ?

**Comment by renatomefidf** [comment]
@eXtreme sadly I have no solution for our case, you may need to do custom returns in each place you need it

**Comment by eXtreme** [comment]
@renatomefidf the easiest thing to do (and the thing I did, cuz it just works) is to write a `kernel.exception` listener and return `JsonResponse` with custom format, but it is outside fosrest.

**Comment by hsz** [comment]
@eXtreme 

I had similar issue and solved it with `ExceptionWrapperHandler`:

```class ExceptionWrapperHandler implements ExceptionWrapperHandlerInterface
{

    /**
     * @param array $data
     *
     * @return ExceptionWrapper
     */
    public function wrap($data)
    {
        if ($data['errors'] instanceof Form) {
            $form = $data['errors'];
            $data['errors'] = [];
            foreach ($form->getErrors(true, true) as $error) {
                $path = $error->getCause()->getConstraint()->errorPath;
                $data['errors'][$path] = $error->getMessage();
            }
        }
        return new ExceptionWrapper($data);
    }

}
```

In my `Controller` I simply return `Form` if it's not valid:

```
class UserController extends FOSRestController
{

    /**
     * @Post("/register")
     *
     * @param Request $request
     * @return array
     */
    public function registerAction(Request $request)
    {
        $form = $this->container->get('fos_user.registration.form');
        /** @var $userManager \FOS\UserBundle\Model\UserManagerInterface */
        $userManager = $this->container->get('fos_user.user_manager');

        $user = $userManager->createUser();
        $form->setData($user);

        $form->handleRequest($request);

        if (!$form->isValid()) {
            return $form;
        }

        $user->setEnabled(true);
        $userManager->updateUser($user);

        return $user;
    }

}
```

**Comment by eXtreme** [comment]
@hsz thanks, I will try it.

**Comment by Ener-Getick** [comment]
Should be fixed by #1358

**Comment by nikunjkabariya** [comment]
Try this class as service. It helped me. https://gist.github.com/Graceas/6505663
